### PR TITLE
User list implementation

### PR DESCRIPTION
**Pull request by tinoqq**
## Summary
user list implementation
## Trello Card
https://trello.com/c/lvxgnVlI

### CODE REVIEW

File: app/controllers/user.js [hunk 1]
```diff
@@ -2,15 +2,43 @@ const User = require('../models').User,
...
   nonExistingUser: 'User does not exist',
   invalidPassword: 'Invalid password',
+  invalidToken: 'Invalid Token',
   emailIsAlreadyRegistered: 'email is already registered'
 };
 exports.badRequestErrorMessages = errorMsgs;
 
+exports.logIn = ({ user }, res, next) => {
+  const email = user.email;
+  User.find({ where: { email } })
```
Comment (molinajulian) [inline app/controllers/user.js]: Send `user.email` directly and remeber return the promise

File: app/controllers/user.js [hunk 2]
```diff
@@ -54,3 +82,29 @@ exports.signUp = ({ user }, res, next) =>
         );
     })
     .catch(next);
+
+exports.usersPerPage = 2;
```
Comment (molinajulian) [inline app/controllers/user.js]: Create a environment var for this

File: app/controllers/user.js [hunk 3]
```diff
@@ -54,3 +82,29 @@ exports.signUp = ({ user }, res, next) =>
...
+  try {
+    const decoded = jwt.decode(req.headers.token);
+  } catch (e) {
+    next(errors.badRequest(errorMsgs.invalidToken));
+    return;
+  }
+  const where = { limit: exports.usersPerPage, offset: 0 };
+  if (req.headers.page) where.offset = req.headers.page * exports.usersPerPage;
+
+  User.findAll(where)
```
Comment (molinajulian) [inline app/controllers/user.js]: Remember always return promises

File: app/controllers/user.js [hunk 4]
```diff
@@ -54,3 +82,29 @@ exports.signUp = ({ user }, res, next) =>
...
+
+exports.listUsers = (req, res, next) => {
+  try {
+    const decoded = jwt.decode(req.headers.token);
+  } catch (e) {
+    next(errors.badRequest(errorMsgs.invalidToken));
+    return;
+  }
+  const where = { limit: exports.usersPerPage, offset: 0 };
+  if (req.headers.page) where.offset = req.headers.page * exports.usersPerPage;
```
Comment (molinajulian) [inline app/controllers/user.js]: Page and limit in query (url, for example endpoint?page=0&limit=0). Use a default value for page and limit

File: app/controllers/user.js [hunk 5]
```diff
@@ -54,3 +82,29 @@ exports.signUp = ({ user }, res, next) =>
...
+
+  User.findAll(where)
+    .then(dbUsers => {
+      const toSendUsers = dbUsers.map(u => {
+        return { firstName: u.firstName, lastName: u.lastName, email: u.email };
+      });
+
+      return res
+        .status(200)
+        .json({ users: toSendUsers })
```
Comment (molinajulian) [inline app/controllers/user.js]: use .send

File: app/controllers/user.js [hunk 6]
```diff
@@ -54,3 +82,29 @@ exports.signUp = ({ user }, res, next) =>
...
+    next(errors.badRequest(errorMsgs.invalidToken));
+    return;
+  }
+  const where = { limit: exports.usersPerPage, offset: 0 };
+  if (req.headers.page) where.offset = req.headers.page * exports.usersPerPage;
+
+  User.findAll(where)
+    .then(dbUsers => {
+      const toSendUsers = dbUsers.map(u => {
+        return { firstName: u.firstName, lastName: u.lastName, email: u.email };
```
Comment (molinajulian) [inline app/controllers/user.js]: Remove brackets and return

File: app/controllers/user.js [hunk 7]
```diff
@@ -54,3 +82,29 @@ exports.signUp = ({ user }, res, next) =>
         );
     })
     .catch(next);
+
+exports.usersPerPage = 2;
+
+exports.listUsers = (req, res, next) => {
+  try {
+    const decoded = jwt.decode(req.headers.token);
```
Comment (molinajulian) [inline app/controllers/user.js]: Why decode token? Remove this. decoded never used

File: test/user.js [hunk 8]
```diff
@@ -1,29 +1,242 @@
...
+            const emailsInPage = userListResponse.body.users.map(u => u.email);
+            expect(emails).to.containSubset(emailsInPage);
+            expect(emailsInPage.length).to.equal(usersPerPage);
+            done();
+          })
+      )
+    );
+  });
+
+  it('it should only return the the amount of users per page', done => {
```
Comment (molinajulian) [inline test/user.js]: Fix wording :D

File: app/controllers/user.js (hunk 5 above)
Comment (molinajulian) [reply app/controllers/user.js]: The response must be pagined

File: test/user.js [hunk 9]
```diff
@@ -1,29 +1,242 @@
...
+      .send()
+      .catch(e => {
+        should.equal(e.response.body.internal_code, errors.BAD_REQUEST);
+        should.equal(e.status, 400);
+        expect(e.response.body.message).to.equal(badRequestErrorMessages.invalidToken);
+        done();
+      });
+  });
+
+  it('should fail because page is empty', done => {
```
Comment (molinajulian) [inline test/user.js]: This test is not correct. If the user dont send page or limit, you must use a default values for page and limit

**Review by molinajulian — CHANGES_REQUESTED** [review]

File: app/controllers/user.js (hunk 7 above)
Comment (tinoqq) [reply app/controllers/user.js]: user list card states: -Se debe estar autenticado para consumir dicho recurso.
though constant decoded is never used (and will be removed), decoded must be performed to compare the token against the api secret

File: app/controllers/user.js (hunk 6 above)
Comment (tinoqq) [reply app/controllers/user.js]: cant map directly to literal object notation because expects code within brackets.

Comment (molinajulian) [reply app/controllers/user.js]: Test this .map(u => ({ new object }))

File: app/controllers/user.js [hunk 10]
```diff
@@ -54,3 +81,30 @@ exports.signUp = ({ user }, res, next) =>
         );
     })
     .catch(next);
+
+exports.listUsers = (req, res, next) => {
+  try {
+    jwt.decode(req.headers.token);
+  } catch (e) {
+    next(errors.badRequest(errorMsgs.invalidToken));
+    return;
```
Comment (molinajulian) [inline app/controllers/user.js:90]: Why this return ?

**Review by molinajulian — CHANGES_REQUESTED** [review]

**Pull request closed without merging**